File: sdk/javascript/src/governance.ts
/**
 * UATP JavaScript/TypeScript SDK - Governance Module
 * 
 * Provides democratic governance participation capabilities for UATP network decisions.
 */

import { Timestamp, UUID } from './types';

export enum ProposalStatus {
  DRAFT = 'draft',
  ACTIVE = 'active',
  VOTING = 'voting',
  PASSED = 'passed',
  REJECTED = 'rejected',
  EXECUTED = 'executed',
  EXPIRED = 'expired'
}

export enum VoteChoice {
  APPROVE = 'approve',
  REJECT = 'reject',
  ABSTAIN = 'abstain'
}

export interface Proposal {
  proposalId: string;
  title: string;
  description: string;
  proposerId: string;
  proposalType: 'constitutional' | 'economic' | 'technical' | 'social';
  status: ProposalStatus;
  votingStart: Timestamp;
  votingEnd: Timestamp;
  requiredQuorum: number;
  requiredMajority: number;

  // Voting results
  votesApprove: number;
  votesReject: number;
  votesAbstain: number;
  totalVotingPower: number;

  // Execution details
  executionData?: Record<string, any>;
  impactAssessment?: Record<string, any>;

  createdAt: Timestamp;
  updatedAt: Timestamp;

  // Computed properties
  readonly totalVotes: number;
  readonly approvalPercentage: number;
  readonly isActive: boolean;
}

export interface Vote {
  voteId: string;
  proposalId: string;
  voterId: string;
  choice: VoteChoice;
  votingPower: number;
  reasoning?: string;
  timestamp: Timestamp;
}

export interface CreateProposalOptions {
  title: string;
  description: string;
  proposalType: 'constitutional' | 'economic' | 'technical' | 'social';
  executionData?: Record<string, any>;
  votingDurationHours?: number;
}

export interface CastVoteOptions {
  proposalId: string;
  choice: VoteChoice | string;
  reasoning?: string;
}

export interface DelegateVotingPowerOptions {
  delegateTo: string;
  amount?: number;
}

/**
 * Client for participating in UATP democratic governance.
 */
export class GovernanceClient {
  private client: any;
  private proposalCache: Map<string, Proposal> = new Map();
  private userVotes: Map<string, Map<string, Vote>> = new Map();

  constructor(client: any) {
    this.client = client;
  }

  /**
   * Get all currently active proposals for voting.
   */
  async getActiveProposals(): Promise<Proposal[]> {
    try {
      const response = await this.client.request('GET', '/api/v1/governance/proposals/active');
      const data = response.data;
      const proposals: Proposal[] = [];

      for (const item of data.proposals || []) {
        const proposal = this.createProposalFromData(item);
        proposals.push(proposal);
        this.proposalCache.set(proposal.proposalId, proposal);
      }

      this.client.emit('active_proposals_retrieved', { count: proposals.length });
      return proposals;

    } catch (error) {
      console.error('Failed to get active proposals:', error);
      return [];
    }
  }

  /**
   * Get a specific proposal by ID.
   */
  async getProposal(proposalId: string): Promise<Proposal | null> {
    // Check cache first
    if (this.proposalCache.has(proposalId)) {
      return this.proposalCache.get(proposalId)!;
    }

    try {
      const response = await this.client.request('GET', `/api/v1/governance/proposals/${proposalId}`);
      const data = response.data;

      const proposal = this.createProposalFromData(data);
      this.proposalCache.set(proposalId, proposal);
      return proposal;

    } catch (error) {
      console.error(`Failed to get proposal ${proposalId}:`, error);
      return null;
    }
  }

  private createProposalFromData(data: any): Proposal {
    const proposal = {
      proposalId: data.proposalId,
      title: data.title,
      description: data.description,
      proposerId: data.proposerId,
      proposalType: data.proposalType,
      status: data.status as ProposalStatus,
      votingStart: data.votingStart,
      votingEnd: data.votingEnd,
      requiredQuorum: data.requiredQuorum,
      requiredMajority: data.requiredMajority,
      votesApprove: data.votesApprove || 0,
      votesReject: data.votesReject || 0,
      votesAbstain: data.votesAbstain || 0,
      totalVotingPower: data.totalVotingPower || 0,
      executionData: data.executionData,
      impactAssessment: data.impactAssessment,
      createdAt: data.createdAt,
      updatedAt: data.updatedAt,

      // Computed properties
      get totalVotes(): number {
        return this.votesApprove + this.votesReject + this.votesAbstain;
      },

      get approvalPercentage(): number {
        if (this.totalVotes === 0) return 0;
        return (this.votesApprove / this.totalVotes) * 100;
      },

      get isActive(): boolean {
        const now = new Date();
        const votingStart = new Date(this.votingStart);
        const votingEnd = new Date(this.votingEnd);
        return (
          this.status === ProposalStatus.VOTING &&
          votingStart <= now &&
          now <= votingEnd
        );
      }
    } as Proposal;

    return proposal;
  }

  /**
   * Cast a vote on a governance proposal.
   */
  async castVote(proposalId: string, choice: VoteChoice | string, reasoning?: string): Promise<any> {
    // Normalize vote choice
    let voteChoice: VoteChoice;
    if (typeof choice === 'string') {
      const normalizedChoice = choice.toLowerCase() as VoteChoice;
      if (!Object.values(VoteChoice).includes(normalizedChoice)) {
        throw new Error(`Invalid vote choice: ${choice}. Must be 'approve', 'reject', or 'abstain'`);
      }
      voteChoice = normalizedChoice;
    } else {
      voteChoice = choice;
    }

    const voteRequest = {
      proposalId,
      choice: voteChoice,
      reasoning,
      timestamp: new Date().toISOString()
    };

    try {
      const response = await this.client.request('POST', '/api/v1/governance/votes', voteRequest);
      const result = response.data;

      // Create vote record
      const vote: Vote = {
        voteId: result.voteId,
        proposalId,
        voterId: result.voterId,
        choice: voteChoice,
        votingPower: result.votingPower,
        reasoning,
        timestamp: new Date().toISOString()
      };

      // Cache user's vote
      if (!this.userVotes.has(result.voterId)) {
        this.userVotes.set(result.voterId, new Map());
      }
      this.userVotes.get(result.voterId)!.set(proposalId, vote);

      // Clear proposal cache to force refresh
      this.proposalCache.delete(proposalId);

      this.client.emit('vote_cast', { proposalId, choice: voteChoice });

      return {
        success: true,
        voteId: result.voteId,
        votingPower: result.votingPower,
        confirmation: result.confirmation,
        vote
      };

    } catch (error) {
      console.error(`Failed to cast vote on proposal ${proposalId}:`, error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        voteId: null
      };
    }
  }

  /**
   * Create a new governance proposal.
   */
  async createProposal(options: CreateProposalOptions): Promise<any> {
    const proposalRequest = {
      title: options.title,
      description: options.description,
      proposalType: options.proposalType,
      executionData: options.executionData,
      votingDurationHours: options.votingDurationHours || 168, // 1 week default
      timestamp: new Date().toISOString()
    };

    try {
      const response = await this.client.request('POST', '/api/v1/governance/proposals', proposalRequest);
      const result = response.data;

      this.client.emit('proposal_created', { proposalId: result.proposalId });
      return result;

    } catch (error) {
      console.error('Failed to create proposal:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
        proposalId: null
      };
    }
  }

  /**
   * Get voting history for a user.
   */
  async getUserVotes(userId: string, limit: number = 50): Promise<Vote[]> {
    try {
      const response = await this.client.request('GET', '/api/v1/governance/votes/user', null, {
        headers: {
          'X-User-ID': userId,
          'X-Limit': limit.toString()
        }
      });

      const data = response.data;
      const votes: Vote[] = [];

      for (const item of data.votes || []) {
        const vote: Vote = {
          voteId: item.voteId,
          proposalId: item.proposalId,
          voterId: item.voterId,
          choice: item.choice as VoteChoice,
          votingPower: item.votingPower,
          reasoning: item.reasoning,
          timestamp: item.timestamp
        };
        votes.push(vote);
      }

      return votes;

    } catch (error) {
      console.error(`Failed to get user votes for ${userId}:`, error);
      return [];
    }
  }

  /**
   * Get overall governance statistics.
   */
  async getGovernanceStats(): Promise<any> {
    try {
      const response = await this.client.request('GET', '/api/v1/governance/stats');
      return response.data;

    } catch (error) {
      console.error('Failed to get governance stats:', error);
      return {
        totalProposals: 0,
        activeProposals: 0,
        totalVoters: 0,
        participationRate: 0,
        proposalsByType: {},
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Get a user's voting power and eligibility.
   */
  async getVotingPower(userId: string): Promise<any> {
    try {
      const response = await this.client.request('GET', '/api/v1/governance/voting-power', null, {
        headers: { 'X-User-ID': userId }
      });

      return response.data;

    } catch (error) {
      console.error(`Failed to get voting power for user ${userId}:`, error);
      return {
        userId,
        votingPower: 0,
        eligible: false,
        requirements: {},
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Delegate voting power to another user.
   */
  async delegateVotingPower(options: DelegateVotingPowerOptions): Promise<any> {
    const delegationRequest = {
      delegateTo: options.delegateTo,
      amount: options.amount,
      timestamp: new Date().toISOString()
    };

    try {
      const response = await this.client.request('POST', '/api/v1/governance/delegate', delegationRequest);
      const result = response.data;

      this.client.emit('voting_power_delegated', { delegateTo: options.delegateTo });
      return result;

    } catch (error) {
      console.error(`Failed to delegate voting power to ${options.delegateTo}:`, error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Get the current constitutional framework of UATP.
   */
  async getConstitutionalFramework(): Promise<any> {
    try {
      const response = await this.client.request('GET', '/api/v1/governance/constitution');
      return response.data;

    } catch (error) {
      console.error('Failed to get constitutional framework:', error);
      return {
        principles: [],
        articles: {},
        amendmentProcess: {},
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  }

  /**
   * Clear all cached governance data.
   */
  clearCache(): void {
    this.proposalCache.clear();
    this.userVotes.clear();
    this.client.emit('governance_cache_cleared', {});
  }
}

File: sdk/python/uatp_sdk/governance.py
"""
UATP Python SDK - Governance Module

Provides democratic governance participation capabilities for UATP network decisions.
"""

import logging
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, List, Optional, Union

logger = logging.getLogger(__name__)


class ProposalStatus(Enum):
    """Status of a governance proposal."""

    DRAFT = "draft"
    ACTIVE = "active"
    VOTING = "voting"
    PASSED = "passed"
    REJECTED = "rejected"
    EXECUTED = "executed"
    EXPIRED = "expired"


class VoteChoice(Enum):
    """Vote choices for governance proposals."""

    APPROVE = "approve"
    REJECT = "reject"
    ABSTAIN = "abstain"


@dataclass
class Proposal:
    """A governance proposal in the UATP network."""

    proposal_id: str
    title: str
    description: str
    proposer_id: str
    proposal_type: str  # "constitutional", "economic", "technical", "social"
    status: ProposalStatus
    voting_start: datetime
    voting_end: datetime
    required_quorum: float
    required_majority: float

    # Voting results
    votes_approve: int = 0
    votes_reject: int = 0
    votes_abstain: int = 0
    total_voting_power: float = 0.0

    # Execution details
    execution_data: Optional[Dict[str, Any]] = None
    impact_assessment: Optional[Dict[str, Any]] = None

    created_at: datetime = None
    updated_at: datetime = None

    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.now(timezone.utc)
        if self.updated_at is None:
            self.updated_at = datetime.now(timezone.utc)

    @property
    def total_votes(self) -> int:
        """Total number of votes cast."""
        return self.votes_approve + self.votes_reject + self.votes_abstain

    @property
    def approval_percentage(self) -> float:
        """Percentage of approve votes."""
        if self.total_votes == 0:
            return 0.0
        return (self.votes_approve / self.total_votes) * 100

    @property
    def is_active(self) -> bool:
        """Check if proposal is currently in voting phase."""
        now = datetime.now(timezone.utc)
        return (
            self.status == ProposalStatus.VOTING
            and self.voting_start <= now <= self.voting_end
        )

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary format."""
        return {
            "proposal_id": self.proposal_id,
            "title": self.title,
            "description": self.description,
            "proposer_id": self.proposer_id,
            "proposal_type": self.proposal_type,
            "status": self.status.value,
            "voting_start": self.voting_start.isoformat(),
            "voting_end": self.voting_end.isoformat(),
            "required_quorum": self.required_quorum,
            "required_majority": self.required_majority,
            "votes_approve": self.votes_approve,
            "votes_reject": self.votes_reject,
            "votes_abstain": self.votes_abstain,
            "total_voting_power": self.total_voting_power,
            "execution_data": self.execution_data,
            "impact_assessment": self.impact_assessment,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


@dataclass
class Vote:
    """A vote cast on a governance proposal."""

    vote_id: str
    proposal_id: str
    voter_id: str
    choice: VoteChoice
    voting_power: float
    reasoning: Optional[str] = None
    timestamp: datetime = None

    def __post_init__(self):
        if self.timestamp is None:
            self.timestamp = datetime.now(timezone.utc)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary format."""
        return {
            "vote_id": self.vote_id,
            "proposal_id": self.proposal_id,
            "voter_id": self.voter_id,
            "choice": self.choice.value,
            "voting_power": self.voting_power,
            "reasoning": self.reasoning,
            "timestamp": self.timestamp.isoformat(),
        }


class GovernanceClient:
    """Client for participating in UATP democratic governance."""

    def __init__(self, client):
        self.client = client
        self.proposal_cache = {}
        self.user_votes = {}
        logger.info(" Governance Client initialized")

    async def get_active_proposals(self) -> List[Proposal]:
        """Get all currently active proposals for voting."""

        try:
            response = await self.client.http_client.get(
                "/api/v1/governance/proposals/active"
            )
            response.raise_for_status()

            data = response.json()
            proposals = []

            for item in data.get("proposals", []):
                proposal = Proposal(
                    proposal_id=item["proposal_id"],
                    title=item["title"],
                    description=item["description"],
                    proposer_id=item["proposer_id"],
                    proposal_type=item["proposal_type"],
                    status=ProposalStatus(item["status"]),
                    voting_start=datetime.fromisoformat(item["voting_start"]),
                    voting_end=datetime.fromisoformat(item["voting_end"]),
                    required_quorum=item["required_quorum"],
                    required_majority=item["required_majority"],
                    votes_approve=item.get("votes_approve", 0),
                    votes_reject=item.get("votes_reject", 0),
                    votes_abstain=item.get("votes_abstain", 0),
                    total_voting_power=item.get("total_voting_power", 0.0),
                    execution_data=item.get("execution_data"),
                    impact_assessment=item.get("impact_assessment"),
                    created_at=datetime.fromisoformat(item["created_at"]),
                    updated_at=datetime.fromisoformat(item["updated_at"]),
                )
                proposals.append(proposal)

                # Cache the proposal
                self.proposal_cache[proposal.proposal_id] = proposal

            logger.info(f"️ Retrieved {len(proposals)} active proposals")
            return proposals

        except Exception as e:
            logger.error(f"[ERROR] Failed to get active proposals: {e}")
            return []

    async def get_proposal(self, proposal_id: str) -> Optional[Proposal]:
        """Get a specific proposal by ID."""

        # Check cache first
        if proposal_id in self.proposal_cache:
            return self.proposal_cache[proposal_id]

        try:
            response = await self.client.http_client.get(
                f"/api/v1/governance/proposals/{proposal_id}"
            )
            response.raise_for_status()

            data = response.json()

            proposal = Proposal(
                proposal_id=data["proposal_id"],
                title=data["title"],
                description=data["description"],
                proposer_id=data["proposer_id"],
                proposal_type=data["proposal_type"],
                status=ProposalStatus(data["status"]),
                voting_start=datetime.fromisoformat(data["voting_start"]),
                voting_end=datetime.fromisoformat(data["voting_end"]),
                required_quorum=data["required_quorum"],
                required_majority=data["required_majority"],
                votes_approve=data.get("votes_approve", 0),
                votes_reject=data.get("votes_reject", 0),
                votes_abstain=data.get("votes_abstain", 0),
                total_voting_power=data.get("total_voting_power", 0.0),
                execution_data=data.get("execution_data"),
                impact_assessment=data.get("impact_assessment"),
                created_at=datetime.fromisoformat(data["created_at"]),
                updated_at=datetime.fromisoformat(data["updated_at"]),
            )

            # Cache the proposal
            self.proposal_cache[proposal_id] = proposal
            return proposal

        except Exception as e:
            logger.error(f"[ERROR] Failed to get proposal {proposal_id}: {e}")
            return None

    async def cast_vote(
        self,
        proposal_id: str,
        choice: Union[str, VoteChoice],
        reasoning: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Cast a vote on a governance proposal.

        Args:
            proposal_id: ID of the proposal to vote on
            choice: Vote choice ("approve", "reject", "abstain")
            reasoning: Optional reasoning for the vote

        Returns:
            Vote confirmation with details
        """

        # Normalize vote choice
        if isinstance(choice, str):
            try:
                choice = VoteChoice(choice.lower())
            except ValueError:
                raise ValueError(
                    f"Invalid vote choice: {choice}. Must be 'approve', 'reject', or 'abstain'"
                )

        vote_request = {
            "proposal_id": proposal_id,
            "choice": choice.value,
            "reasoning": reasoning,
            "timestamp": datetime.utcnow().isoformat(),
        }

        try:
            response = await self.client.http_client.post(
                "/api/v1/governance/votes", json=vote_request
            )
            response.raise_for_status()

            result = response.json()

            # Create vote record
            vote = Vote(
                vote_id=result["vote_id"],
                proposal_id=proposal_id,
                voter_id=result["voter_id"],
                choice=choice,
                voting_power=result["voting_power"],
                reasoning=reasoning,
            )

            # Cache user's vote
            if result["voter_id"] not in self.user_votes:
                self.user_votes[result["voter_id"]] = {}
            self.user_votes[result["voter_id"]][proposal_id] = vote

            # Clear proposal cache to force refresh
            if proposal_id in self.proposal_cache:
                del self.proposal_cache[proposal_id]

            logger.info(f"️ Vote cast on proposal {proposal_id}: {choice.value}")
            return {
                "success": True,
                "vote_id": result["vote_id"],
                "voting_power": result["voting_power"],
                "confirmation": result.get("confirmation"),
                "vote": vote.to_dict(),
            }

        except Exception as e:
            logger.error(f"[ERROR] Failed to cast vote on proposal {proposal_id}: {e}")
            return {"success": False, "error": str(e), "vote_id": None}

    async def create_proposal(
        self,
        title: str,
        description: str,
        proposal_type: str,
        execution_data: Optional[Dict[str, Any]] = None,
        voting_duration_hours: int = 168,  # 1 week default
    ) -> Dict[str, Any]:
        """
        Create a new governance proposal.

        Args:
            title: Proposal title
            description: Detailed description
            proposal_type: "constitutional", "economic", "technical", "social"
            execution_data: Data needed for proposal execution
            voting_duration_hours: How long voting should remain open

        Returns:
            Created proposal information
        """

        proposal_request = {
            "title": title,
            "description": description,
            "proposal_type": proposal_type,
            "execution_data": execution_data,
            "voting_duration_hours": voting_duration_hours,
            "timestamp": datetime.utcnow().isoformat(),
        }

        try:
            response = await self.client.http_client.post(
                "/api/v1/governance/proposals", json=proposal_request
            )
            response.raise_for_status()

            result = response.json()
            logger.info(f" Created proposal: {result['proposal_id']}")
            return result

        except Exception as e:
            logger.error(f"[ERROR] Failed to create proposal: {e}")
            return {"success": False, "error": str(e), "proposal_id": None}

    async def get_user_votes(self, user_id: str, limit: int = 50) -> List[Vote]:
        """Get voting history for a user."""

        try:
            params = {"user_id": user_id, "limit": limit}

            response = await self.client.http_client.get(
                "/api/v1/governance/votes/user", params=params
            )
            response.raise_for_status()

            data = response.json()
            votes = []

            for item in data.get("votes", []):
                vote = Vote(
                    vote_id=item["vote_id"],
                    proposal_id=item["proposal_id"],
                    voter_id=item["voter_id"],
                    choice=VoteChoice(item["choice"]),
                    voting_power=item["voting_power"],
                    reasoning=item.get("reasoning"),
                    timestamp=datetime.fromisoformat(item["timestamp"]),
                )
                votes.append(vote)

            logger.info(f" Retrieved {len(votes)} votes for user {user_id}")
            return votes

        except Exception as e:
            logger.error(f"[ERROR] Failed to get user votes for {user_id}: {e}")
            return []

    async def get_governance_stats(self) -> Dict[str, Any]:
        """Get overall governance statistics."""

        try:
            response = await self.client.http_client.get("/api/v1/governance/stats")
            response.raise_for_status()

            return response.json()

        except Exception as e:
            logger.error(f"[ERROR] Failed to get governance stats: {e}")
            return {
                "total_proposals": 0,
                "active_proposals": 0,
                "total_voters": 0,
                "participation_rate": 0.0,
                "proposals_by_type": {},
                "error": str(e),
            }

    async def get_voting_power(self, user_id: str) -> Dict[str, Any]:
        """Get a user's voting power and eligibility."""

        try:
            params = {"user_id": user_id}
            response = await self.client.http_client.get(
                "/api/v1/governance/voting-power", params=params
            )
            response.raise_for_status()

            return response.json()

        except Exception as e:
            logger.error(f"[ERROR] Failed to get voting power for user {user_id}: {e}")
            return {
                "user_id": user_id,
                "voting_power": 0.0,
                "eligible": False,
                "requirements": {},
                "error": str(e),
            }

    async def delegate_voting_power(
        self, delegate_to: str, amount: Optional[float] = None
    ) -> Dict[str, Any]:
        """
        Delegate voting power to another user.

        Args:
            delegate_to: User ID to delegate to
            amount: Amount of voting power to delegate (None for all)

        Returns:
            Delegation confirmation
        """

        delegation_request = {
            "delegate_to": delegate_to,
            "amount": amount,
            "timestamp": datetime.utcnow().isoformat(),
        }

        try:
            response = await self.client.http_client.post(
                "/api/v1/governance/delegate", json=delegation_request
            )
            response.raise_for_status()

            result = response.json()
            logger.info(f" Delegated voting power to {delegate_to}")
            return result

        except Exception as e:
            logger.error(
                f"[ERROR] Failed to delegate voting power to {delegate_to}: {e}"
            )
            return {"success": False, "error": str(e)}

    async def get_constitutional_framework(self) -> Dict[str, Any]:
        """Get the current constitutional framework of UATP."""

        try:
            response = await self.client.http_client.get(
                "/api/v1/governance/constitution"
            )
            response.raise_for_status()

            return response.json()

        except Exception as e:
            logger.error(f"[ERROR] Failed to get constitutional framework: {e}")
            return {
                "principles": [],
                "articles": {},
                "amendment_process": {},
                "error": str(e),
            }

    def clear_cache(self):
        """Clear all cached governance data."""
        self.proposal_cache.clear()
        self.user_votes.clear()
        logger.info(" Governance cache cleared")
